source/transfer_list.rst

```diff
@@ -174,13 +198,6 @@ The *tag_id* ranges are described in :numref:`tab_tag_id_ranges`.
 Standard transfer entries
 -------------------------
 
-The TEs have a *tag_id* in the {0, ..., 0xf_ffff} set. Both
-the tag_id of a standard entry as well as the entry layout
-must be defined in this specification before being used.
-New entries are expected to have a simple layout. Complex
-data should be represented in a self-describing data
-structure, such as the FDT [DT]_.
```

Comment:
Er, I don't agree to deleting this. The FDT is where more complex things should go. The TL is not a way to replicate TF-A's bl_param thing, nor coreboot's tag thing, nor Linux's atags thing, except as needed in early stages when memory is scarce or the CPU is running slowly. Those structures should move to FDT, at least for later states where there is plenty of memory.

Reply:
Can we accommodate this last sentence in the policy above (e.g. around line 155)? The language can be softened if you prefer, or to refer to other complex data representations (e.g. ACPI/HOB).

Reply:
Well... I still don't think that this specification should express any preference for other wrapped data structures, or even the concept that certain data needs to be wrapped in other structures in general. That goes back to that whole argument that I don't think it makes sense to invent a new data representation whose only purpose it is to wrap existing representations, we could just continue using the existing representations in that case. I'm fine with saying that users of this specification may use wrapped data structures and keep stuff in there as they see fit, but I don't think there should be explicit bias towards that.

I understand Simon's desire to discourage "complex" data structures, but I think one reasonable solution to that is to split everything up into individual tags until it is no longer "complex" -- that's what I was trying to express with the "simple C structure" and "single logical concept" text further up. I think an implementation that chooses to keep all information in a large number of small, distinct TEs like that should be just as valid.

I don't really understand the "dynamic through the boot flow" part -- why should we discourage using raw TEs to store information that may change? I don't see why an embedded FDT or ACPI table would be a better place to store that (in fact, updating something inside a large, embedded structure like that would likely be a lot more painful than updating a small, single-issue TE).

Reply:
I think my comment here was lost...

P1: I believe that device tree is the best structure to use for new things, unless they are just a simple C structure and need to be handled before using FDT is practical. That discussion is also being hashed out in the coreboot mailing list, I suppose.

P2: Yes, splitting things into smaller data structures is one approach. I think we would need a particular reason for doing that, such as

a) it is pre-existing, like with coreboot
b) it is needed when the system is limited on memory / performance

otherwise we'll end up with effectively inventing out own schema which FDT would handle better.

P3: I don't see an absolute problem with TEs changing, but it is a sign of misuse. It is perfectly reasonable to add to a TPM log (say) in consecutive firmware phases, but it modifying various TEs in a boot phase suggests that they might have been poorly defined.

For example, it might make more sense to define a TE that indicates what is known about a particular thing, e.g. the video console, with that being immutable. Then once the boot proceeds, leave that alone (as a lowest-common-denominator description) and put new / higher-level info in an FDT. 

Of course there are likely some cases (besides growing logs / tables) where updating a TE is sensible, so I don't think we should enforce a blanket rule.

Reply:
I don't feel too strongly whether we add some more text here or not; I was just trying to suggest some words that accommodate what IMO are both valid use-cases. By "dynamic through the boot flow" I was meaning structures that are passed/modified through more than 2 components like the kernel FDT, but I agree the wording I suggested isn't great. I guess what I was trying to say is that if a later firmware/OS component *requires* the passing of data in an existing format like FDT, then it may be better for earlier firmware to use that format too (if it operates on the data). rather than creating dedicated TEs that are later converted into the existing format.
I accept it's valid to dynamically modify distinct TEs.

Reply:
For now, please drop this change. There is too much going on in this PR already!

Once we get this PR in, let's do a new one with just this topic...eg.g. I want to make sure people don't do things like:

TE1: Wibble header
TE2: wibble record 1
TE3: wibble record 2
TE4: wibble record 3

It is OK to split structures but we need to think about the terminology.

So please drop it for now.

Reply:
I can drop the last patch from this PR once we think the others are ready for merge. Unfortunately GitHub doesn't allow stacks of dependent PRs, so once I do that we wouldn't have a place to discuss this part any more until the rest have landed.

Regarding your latest example, I totally agree that there shouldn't be any dependencies between TEs or any meaning implied by their order or anything like that, if that's what you were trying to illustrate there. Every TE should be totally independent and stand alone, and it never makes sense that one TE is the "header" for another TE. I can add more explicit language in that direction with my next update.

However, I think it should be legal if an unordered group of entities is represented by one TE per element (with them all having the same or related tag IDs). In fact, we have an example of that in the current version of the spec already with the HOB block TE. So if you equate "wibble" with "HOB block" and leave out the header TE, it would be legal to have

TE2: HOB block 1
TE3: HOB block 2
TE4: HOB block 3

(with the understanding that the order between the HOB blocks is arbitrary and there may be other TEs with different tag IDs in between those entries).

Reply:
Agreed, sounds good.